### PR TITLE
Add recent posts API endpoint

### PR DESCRIPTION
**Pull request by rosszurowski**
This commit introduces a recent posts API endpoint which returns details about the most recent blog posts for use in throughout the site. This approach is easier to integrate than parsing the RSS feed, and lets us provide additional details which aren't currently included in the RSS feed.

### CODE REVIEW

File: pages/api/recent-posts.ts [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+  const siteMap = await getSiteMap()
+  const ttl = 24 * 60 * 60 // 24 hours
+  const data: Response = {
+    posts: []
+  }
+
+  const limit = 5
+
+  for (const pagePath of Object.keys(siteMap.canonicalPageMap)) {
+    const pageId = siteMap.canonicalPageMap[pagePath]
```
Comment (rosszurowski) [inline pages/api/recent-posts.ts:47]: A chunk of this code is duplicated from `feed.tsx`. We could extract it out into common utilities, but I figured a little duplication for these different use-cases is ok.

File: pages/api/recent-posts.ts [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+      url,
+      authors: authorList
+        .split(',')
+        // Remove any uses of the word "and", along with any extra whitespace.
+        .map((s) => s.replace(/\band\b/g, '').trim())
+    })
+  }
+
+  // Sort posts from newest to oldest.
+  data.posts.sort((a, b) => b.published - a.published).slice(0, limit)
```
Comment (rosszurowski) [inline pages/api/recent-posts.ts:103]: It's a little dirty to load all the data and then limit it here, but I couldn't figure out how to do this via the pageMap.

File: pages/api/recent-posts.ts [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+      continue
+    }
+
+    const isPublished = getPageProperty<number>('Published', block, recordMap)
+    if (!isPublished) {
+      continue
+    }
+
+    const title = getBlockTitle(block, recordMap) || config.name
+    const description =
```
Comment (rosszurowski) [inline pages/api/recent-posts.ts:70]: I noticed pretty much every blog post is missing the description, so they all fall back to the site-wide description, which probably isn't great for SEO.

I started to explore extracting an excerpt, but Notion's block-based approach makes it kind of tricky. I'd recommend including a short excerpt in this Description field in Notion. I'm happy to provide backfill data for all of these if it's helpful. At the very least, we should do this for the most recent posts which appear on the site.

For example:

- **What Can Voltage Structures…**: Knowing the structure of the grid—how lines interconnect and what phases loads are on—is vital for efficient grid maintenance and operations, informing applications ranging from fault localization to phase balancing. 
- **Estimating SAIDI…**: How can we estimate SAIDI (System Average Interruption Duration Index)—the average power outage duration experienced across all customers served—from PowerWatch sensor measurements of only a subset of customers?
- **A Clustering Algorithm**: nLine installs power sensors at outlets in homes, small businesses, and social infrastructure. How do we estimate the extent of a grid outage from individual sensor reports?

**Comment by rosszurowski** [comment]
Looks like the deployment failed, but I'm not sure why. I noticed locally some errors because [the CSS file was missing a closing brace](https://github.com/nline/nline-blog/blob/22c0fcfc671e8b3629f5401d06748c65001dd437/styles/nline.css#L60-L62), which might be the cause.

File: pages/api/recent-posts.ts (hunk 3 above)
Comment (jacksongoode) [reply pages/api/recent-posts.ts:70]: Looking at this right now, I've wanted to add descriptions for these posts from Notion's side.

Comment (jacksongoode) [reply pages/api/recent-posts.ts:70]: Ok I added a description to each. They are various lengths but I imagine some overflow might take care of that?

File: pages/api/recent-posts.ts [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+  posts: BlogPost[]
+}
+
+/**
+ * This handler exposes recent blog posts as a JSON object for displaying post
+ * previews on the rest of the site.
+ */
+export default async function (req: NextApiRequest, res: NextApiResponse) {
+  const siteMap = await getSiteMap()
+  const ttl = 24 * 60 * 60 // 24 hours
```
Comment (jacksongoode) [inline pages/api/recent-posts.ts:39]: I would say we could even bump this to like 3 or 4 days?

File: pages/api/recent-posts.ts [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    const socialImageUrl = getSocialImageUrl(pageId)
+    const authorList =
+      getPageProperty<string>('Authors', block, recordMap) || config.author
+    const published = getPageProperty<number>('Date', block, recordMap)
+
+    // By default, social image URLs have a big title in front. We add an extra
+    // query parameter to signal that we want to use the image as-is.
+    const revisedImageUrl = new URL(socialImageUrl)
+    revisedImageUrl.searchParams.set('text', 'false')
+    const coverImageUrl = revisedImageUrl.toString()
```
Comment (jacksongoode) [inline pages/api/recent-posts.ts:83]: I don't think we're actually getting the right image url. In most cases it's blank? This is an existing issue. I'll see if there's a workaround.

Comment (jacksongoode) [reply pages/api/recent-posts.ts:83]: I switched the way the socialImage is generated to first get the cover image. I'm not sure why it couldn't get the first image block as it does for the card previews on, but I think this is actually better.

File: styles/nline.css [hunk 6]
```diff
@@ -39,26 +39,25 @@ body {
...
-  -webkit-box-shadow: inset 0 0 6px rgba(0,0,0,.3);
+  -webkit-box-shadow: inset 0 0 6px rgba(0, 0, 0, 0.3);
   background-color: var(--fg-color-1);
 }
 
-::-webkit-scrollbar-track {           
-    background-color: var(--bg-color);
+::-webkit-scrollbar-track {
+  background-color: var(--bg-color);
+}
```
Comment (jacksongoode) [inline styles/nline.css:60]: This was added to upstream but I can't rebase on your main so I just did it here as well.

**Review by jacksongoode — APPROVED** [review]
This looks good. We can always make changes later but the endpoint seems to work.

File: pages/api/recent-posts.ts (hunk 5 above)
Comment (rosszurowski) [reply pages/api/recent-posts.ts:83]: Nice! I noticed it was just some of the newer posts that had them, but if you're able to get it working for more, that's great. We're just pulling the latest few posts anyways, so it didn't seem like an issue.

**Pull request merged**